Sources/MongoSwift/ConnectionPool.swift

```diff
@@ -67,38 +93,80 @@ internal class ConnectionPool {
+        }
+
+        // continually loop and wait to get all connections back before destroying the pool. release the lock on each
+        // iteration to allow other methods to acquire the lock.
+        var done = false
+        while !done {
+            try self.stateLock.withLock {
+                if self.connCount == 0 {
+                    switch self.state {
+                    case .open, .closed:
+                        throw InternalError(message: "ConnectionPool in unexpected state \(self.state) during close()")
+                    case let .closing(pool):
+                        mongoc_client_pool_destroy(pool)
+                        self.state = .closed
+                    }
+                    done = true
+                }
+            }
         }
-        self.state = .closed
     }
 
     /// Checks out a connection. This connection will return itself to the pool when its reference count drops to 0.
-    /// This method will block until a connection is available.
+    /// This method will block until a connection is available. Throws an error if the pool is in the process of
+    /// closing or has finished closing.
     internal func checkOut() throws -> Connection {
-        switch self.state {
-        case let .open(pool):
-            return Connection(clientHandle: mongoc_client_pool_pop(pool), pool: self)
-        case .closed:
-            throw InternalError(message: "ConnectionPool was already closed")
+        try self.stateLock.withLock {
+            switch self.state {
+            case let .open(pool):
+                self.connCount += 1
+                return Connection(clientHandle: mongoc_client_pool_pop(pool), pool: self)
+            case .closing, .closed:
+                throw Self.PoolClosedError
+            }
         }
     }
 
-    /// Checks out a connection from the pool, or returns `nil` if none are currently available.
+    /// Checks out a connection from the pool, or returns `nil` if none are currently available. Throws an error if the
+    /// pool is not open.
     internal func tryCheckOut() throws -> Connection? {
-        switch self.state {
-        case let .open(pool):
-            guard let handle = mongoc_client_pool_try_pop(pool) else {
-                return nil
+        try self.stateLock.withLock {
+            switch self.state {
+            case let .open(pool):
+                guard let handle = mongoc_client_pool_try_pop(pool) else {
+                    return nil
+                }
+                self.connCount += 1
+                return Connection(clientHandle: handle, pool: self)
+            case .closing, .closed:
+                throw Self.PoolClosedError
+            }
+        }
+    }
+
+    /// Checks  a connection into the pool. Accepts the connection if the pool is still open or in the process of
+    /// closing; throws an error if the pool has already finished closing.
+    fileprivate func checkIn(_ connection: Connection) throws {
```

Comment:
In the success case, there wouldn't really be any behavioral difference. The main benefit there would just be that the blocking code is done more explicitly, so there isn't a possibility we do it by accident. 

In the error case (i.e. when the user forgets to call `kill` or `end`), the `deinit` approach prevents a memory leak but may block the event loop, whereas if we only explicitly check in connections, the user would get a memory leak but not stall the event loop. It's not really a big deal either way though, since ideally users will be closing their resources.

Anyways, this is pretty minor so either way is fine. I figured it was worth consideration though.

Reply:
ah yeah. I don't really know which bad outcome is worse. I've opened [SWIFT-825](https://jira.mongodb.org/browse/SWIFT-825) so we can discuss in triage.